Java-FlightServer/src/main/java/ibm/com/example/RelayFlightServer.java

```diff
@@ -0,0 +1,149 @@
+                this.listener = listener;
+            }
+
+            @Override
+            public WaitResult waitForListener(long timeout) {
+                while (!listener.isReady() && !listener.isCancelled()) {
+                    // busy wait
+                }
+                return WaitResult.READY;
+            }
+        }, false);
+    }
+
+    public RelayFlightServer(BufferAllocator incomingAllocator, Location location, BackpressureStrategy bpStrategy,
+                             boolean isNonBlocking) {
+        this.allocator = incomingAllocator.newChildAllocator("flight-server", 0, Long.MAX_VALUE);
+        this.location = location;
+        Producer producer = new Producer(bpStrategy);
+        this.flightServer = FlightServer.builder(this.allocator, location, producer).build();
+        this.isNonBlocking = isNonBlocking;
+        this.client = FlightClient.builder()
+                .allocator(allocator)
+                .location(Location.forGrpcInsecure("localhost", 12233))
+                .build();
+    }
+
+    public Location getLocation() {
+        return location;
+    }
+
+    public void start() throws IOException {
+        flightServer.start();
+    }
+
+    public void awaitTermination() throws InterruptedException {
+        flightServer.awaitTermination();
+    }
+
+    /**
+     * Main method starts the server listening to localhost:12232.
+     */
+    public static void main(String[] args) throws Exception {
+        final BufferAllocator a = new RootAllocator(Long.MAX_VALUE);
+        final RelayFlightServer efs = new RelayFlightServer(a, Location.forGrpcInsecure("localhost", 12232));
+        efs.start();
+        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+            try {
+                System.out.println("\nExiting...");
+                AutoCloseables.close(efs, a);
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }));
+        efs.awaitTermination();
+    }
+
+    private final class Producer extends NoOpFlightProducer {
+        private final BackpressureStrategy bpStrategy;
+
+        private Producer(BackpressureStrategy bpStrategy) {
+            this.bpStrategy = bpStrategy;
+        }
+
+        @Override
+        public void getStream(CallContext context, Ticket ticket,
+                              ServerStreamListener listener) {
+            bpStrategy.register(listener);
+            FlightStream s = client.getStream(ticket);
+            final Runnable loadData = () -> {
+                VectorSchemaRoot root;
+                listener.setUseZeroCopy(true);
+                boolean first = true;
+                while (s.next()) {
+                    root = s.getRoot();
+                    if (first) {
+                        listener.start(root);
+                        first = false;
+                    }
+                    bpStrategy.waitForListener(0);
```

Comment:
busy wait for listener to be ready

Reply:
In the examples (cf the perf test) they don't use it. Did you check what it adds?

Reply:
@roee88: without waiting for the listener, the relay server sometimes crashes. That is not the case for the example server. Let's discuss and try to figure it out.